jaxtronomy/LensModel/Solver/lens_equation_solver.py

```diff
@@ -0,0 +1,773 @@
+
+        :param sourcePos_x: source position in units of angle
+        :param sourcePos_y: source position in units of angle
+        :param kwargs_lens: lens model parameters as keyword arguments
+        :param min_distance: minimum separation to consider for two images in units of
+            angle
+        :param search_window: window size to be considered by the solver. Will not find
+            image position outside this window
+        :param precision_limit: required precision in the lens equation solver (in units
+            of angle in the source plane).
+        :param num_iter_max: maximum iteration of lens-source mapping conducted by
+            solver to match the required precision
+        :param arrival_time_sort: bool, if True, sorts image position in arrival time
+            (first arrival photon first listed)
+        :param initial_guess_cut: bool, if True, cuts initial local minima selected by
+            the grid search based on distance criteria from the source position
+        :param verbose: bool, if True, prints some useful information for the user
+        :param x_center: float, center of the window to search for point sources
+        :param y_center: float, center of the window to search for point sources
+        :param num_random: int, number of random positions within the search window to
+            be added to be starting positions for the gradient decent solver
+        :param magnification_limit: None or float, if set will only return image
+            positions that have an abs(magnification) larger than this number
+        :returns: (exact) angular position of (multiple) images ra_pos, dec_pos in units
+            of angle
+        :raises: AttributeError, KeyError
+        """
+        # find pixels in the image plane possibly hosting a solution of the lens equation, related source distances and
+        # pixel width
+        x_mins, y_mins, delta_map, pixel_width = self.candidate_solutions(
+            sourcePos_x,
+            sourcePos_y,
+            kwargs_lens,
+            min_distance,
+            search_window,
+            x_center,
+            y_center,
+        )
+        if verbose:
+            print(
+                "There are %s regions identified that could contain a solution of the lens equation with"
+                "coordinates %s and %s " % (len(x_mins), x_mins, y_mins)
+            )
+        if initial_guess_cut and len(x_mins > 0):
+            mag = np.abs(self.lensModel.magnification(x_mins, y_mins, kwargs_lens))
+            mag[mag < 1] = 1
+            x_mins = x_mins[delta_map <= min_distance * mag * 5]
+            y_mins = y_mins[delta_map <= min_distance * mag * 5]
+            if verbose:
+                print(
+                    "The number of regions that meet the plausibility criteria are %s"
+                    % len(x_mins)
+                )
+        if len(x_mins) < 1:
+            return x_mins, y_mins
+        x_mins = np.append(
```

Comment:
is this supported with JAX? Might also not be needed in lenstronomy either way (adding random candidate solutions from where to do gradient decent)

Reply:
I left image_position_lenstronomy and candidate_solutions unjaxified, only the gradient descent portion of the algorithm has been jaxified. I have jit-decorated the computation-heavy functions so that users can still take advantage of jax even without using jax.jit on the overall solver (e.g. if you use `image_position_from_source`, you will still get speed up without using `jax.jit(image_position_from_source)` since the subfunctions have been jit-decorated). Due to the restrictions of writing in jax, I do not think it is possible to jaxify everything.